File: ployglots/mta-ployglot-pkg-figma-api/packages/py/figma_api/_fetch_client.py
"""Pluggable FetchClient contract — the only transport surface
FigmaClient depends on. The default implementation wraps the
``polyglot-fetch-http-client`` ``AsyncClient``; callers can swap in
their own for BYO compositions.
"""

from __future__ import annotations

from typing import Any, Protocol

from fetch_http_client import APIKeyAuth, AsyncClient

from ._config import FigmaConfig, resolve_figma_config
from ._proxy import build_proxy
from ._retry import build_figma_retry_config
from ._version import __version__


class FetchClient(Protocol):
    async def get(self, path: str, **kwargs: Any) -> Any: ...
    async def post(self, path: str, **kwargs: Any) -> Any: ...
    async def put(self, path: str, **kwargs: Any) -> Any: ...
    async def delete(self, path: str, **kwargs: Any) -> Any: ...
    async def patch(self, path: str, **kwargs: Any) -> Any: ...
    async def aclose(self) -> None: ...


def create_default_fetch_client(config: FigmaConfig) -> AsyncClient:
    """Build the default Figma FetchClient from a resolved config."""
    proxy = build_proxy(config.proxy)

    # `retry=None` passed to FigmaClient means "use Figma defaults".
    # `retry=False` means "disable". Pass-through dict gets merged.
    retry_input = config.retry if config.retry is not None else {}
    retry = build_figma_retry_config(
        retry_input, force_overwrite=config.force_overwrite_retry
    )

    kwargs: dict[str, Any] = {
        "base_url": config.host,
        "auth": APIKeyAuth(config.token, "X-Figma-Token"),
        "timeout": config.timeout_ms / 1000.0,
        "headers": {
            "accept": "application/json",
            "user-agent": f"polyglot-figma-api/{__version__}",
            **config.default_headers,
        },
    }
    if proxy is not None:
        kwargs["proxy"] = proxy
    if retry is not None:
        kwargs["retry"] = retry
    return AsyncClient(**kwargs)


def fetch_client_from_polyglot(inner: AsyncClient) -> AsyncClient:
    """Helper for the BYO pattern — returns the AsyncClient verbatim.

    Kept as a named function so callers can be explicit about intent and
    so the twin surface matches ``fetchClientFromPolyglot`` on the ts
    side.
    """
    return inner


def create_figma_fetch_client(**kwargs: Any) -> AsyncClient:
    """Build a ready-to-use default FetchClient from config inputs."""
    return create_default_fetch_client(resolve_figma_config(**kwargs))

File: ployglots/mta-ployglot-pkg-figma-api/packages/ts/src/fetch-client.ts
// @ts-nocheck
/**
 * Pluggable `FetchClient` contract — the only transport surface
 * FigmaClient depends on. The default implementation wraps
 * `@polyglot/fetch-http-client`'s `AsyncClient`; callers can swap in
 * their own `AsyncClient` (or any object conforming to this interface)
 * for BYO compositions.
 */

import { AsyncClient, APIKeyAuth, type Response } from '@polyglot/fetch-http-client';

import { resolveFigmaConfig, type FigmaConfig, type FigmaConfigInput } from './config.js';
import { buildProxy } from './proxy.js';
import { buildFigmaRetryConfig } from './retry.js';

export interface FetchRequestOptions {
  headers?: Record<string, string>;
  params?: Record<string, string | number | boolean | undefined>;
  body?: unknown;
  signal?: AbortSignal;
  timeout?: number;
}

export interface FetchClient {
  get(path: string, options?: FetchRequestOptions): Promise<Response>;
  post(path: string, options?: FetchRequestOptions): Promise<Response>;
  put(path: string, options?: FetchRequestOptions): Promise<Response>;
  delete(path: string, options?: FetchRequestOptions): Promise<Response>;
  patch(path: string, options?: FetchRequestOptions): Promise<Response>;
  close(): Promise<void>;
}

/**
 * Build the default Figma `FetchClient` from a resolved config.
 *
 * This is the glue between `@polyglot/figma-api` and
 * `@polyglot/fetch-http-client` — swap or subclass it to change
 * transport behavior. The configured AsyncClient already satisfies
 * `FetchClient` (verb methods + close); we return it directly.
 */
export function createDefaultFetchClient(config: FigmaConfig): FetchClient {
  const proxy = buildProxy(config.proxy);
  const retry = buildFigmaRetryConfig(config.retry, {
    forceOverwrite: config.forceOverwriteRetry,
  });
  const client = new AsyncClient({
    baseUrl: config.host,
    auth: new APIKeyAuth(config.token, 'X-Figma-Token'),
    timeout: config.timeoutMs,
    headers: {
      accept: 'application/json',
      'user-agent': `polyglot-figma-api/${process.env.npm_package_version ?? '0.1.0'}`,
      ...config.defaultHeaders,
    },
    ...(proxy ? { proxy } : {}),
    ...(retry === false ? { retry: false } : { retry }),
  });
  return client as unknown as FetchClient;
}

/**
 * Helper for callers composing their own outer `AsyncClient` and
 * wanting a `FetchClient` wrapping it — the "BYO adapter" in the
 * plan's F03/04 story. Use when you need outer retry / circuit
 * breaker / caching on top of the Figma client.
 *
 * @example
 *   const outer = new AsyncClient({
 *     baseUrl: 'https://api.figma.com',
 *     auth: new APIKeyAuth(token, 'X-Figma-Token'),
 *     retry: { maxRetries: 5 },
 *   });
 *   const client = new FigmaClient({
 *     token,
 *     fetchClient: fetchClientFromPolyglot(outer),
 *   });
 */
export function fetchClientFromPolyglot(inner: AsyncClient): FetchClient {
  return inner as unknown as FetchClient;
}

/**
 * Build a ready-to-use default `FetchClient` from a config input —
 * equivalent to `new FigmaClient({...}).fetchClient`, but exposed for
 * callers who want to share one transport across multiple clients.
 */
export function createFigmaFetchClient(input: FigmaConfigInput = {}): FetchClient {
  return createDefaultFetchClient(resolveFigmaConfig(input));
}
